va/va.go

```diff
@@ -639,23 +640,34 @@ func (va *ValidationAuthorityImpl) validateChallenge(ctx context.Context, identi
 	return nil, probs.Malformed("invalid challenge type %s", challenge.Type)
 }
 
-func (va *ValidationAuthorityImpl) performRemoteValidation(ctx context.Context, domain string, challenge core.Challenge, authz core.Authorization, result chan *probs.ProblemDetails) {
-	s := va.clk.Now()
-	errors := make(chan error, len(va.remoteVAs))
+// performRemoteValidation calls `PerformValidation` for each of the configured
+// remoteVAs in a random order. The provided `results` chan should have an equal
+// size to the number of remote VAs. The validations will be peformed in
+// separate go-routines. If the result `error` from a remote
+// `PerformValidation` RPC is nil or a nil `ProblemDetails` instance it is
+// written directly to the `results` chan. If the err is a cancelled error it is
+// treated as a nil error. Otherwise the error/problem is written to the results
+// channel as-is.
+func (va *ValidationAuthorityImpl) performRemoteValidation(
+	ctx context.Context,
+	domain string,
+	challenge core.Challenge,
+	authz core.Authorization,
+	results chan error) {
 	for _, i := range rand.Perm(len(va.remoteVAs)) {
 		remoteVA := va.remoteVAs[i]
-		go func(rva RemoteVA) {
+		go func(rva RemoteVA, index int) {
 			_, err := rva.PerformValidation(ctx, domain, challenge, authz)
 			if err != nil {
 				// returned error can be a nil *probs.ProblemDetails which breaks the
 				// err != nil check so do a slightly more complicated unwrap check to
 				// make sure we don't choke on that.
-				if p, ok := err.(*probs.ProblemDetails); ok || p != nil {
+				if p, ok := err.(*probs.ProblemDetails); ok && p != (*probs.ProblemDetails)(nil) {
 					// If the non-nil err was a non-nil *probs.ProblemDetails then we can
 					// log it at an info level. It's a normal non-success validation
 					// result and the remote VA will have logged more detail.
 					va.log.Infof("Remote VA %q.PerformValidation returned problem: %s", rva.Addresses, err)
-				} else if ok && p == nil {
+				} else if ok && p == (*probs.ProblemDetails)(nil) {
```

Comment:
Side note: did we ever figure out what was causing this? Do we have a ticket to fix it?

Reply:
Yeah, I reopened #2254 to track it and will have a small PR addressing this one specific case shortly.

Reply:
Cool, could we add a TODO here referencing that and noting that we should be able to get rid of this confusion at some point.

Reply:
Added a TODO in 67e7b27